src/java.base/share/classes/java/util/stream/Stream.java

```diff
@@ -1051,6 +1053,46 @@ <U> U reduce(U identity,
                  BiFunction<U, ? super T, U> accumulator,
                  BinaryOperator<U> combiner);
 
+    /**
+     * Returns a stream consisting of the results of applying the given
+     * gatherer to the elements of this stream.
+     *
+     * <p>This is an <a href="package-summary.html#Extensibility">extension point</a>
+     * for <a href="package-summary.html#StreamOps">intermediate operations</a>.
+     *
+     * <p>When executed in parallel, multiple intermediate results may be
+     * instantiated, populated, and merged so as to maintain isolation of
+     * mutable data structures.  Therefore, even when executed in parallel
+     * with non-thread-safe data structures (such as {@code ArrayList}), no
+     * additional synchronization is needed for a parallel reduction.
+     *
+     * <p>Implementations are allowed, but not required, to detect consecutive
+     * invocations and compose them into a single, fused, operation. This would
+     * make the first expression below behave like the second:
+     *
+     * <pre>{@code
+     *     var stream1 = Stream.of(...).gather(gatherer1).gather(gatherer2);
+     *     var stream2 = Stream.of(...).gather(gatherer1.andThen(gatherer2));
+     * }</pre>
+     *
+     * @implSpec The implementation in this interface returns a Stream produced as if by the following:
+     * <pre>{@code
+     * StreamSupport.stream(spliterator(), isParallel()).gather(gatherer)
+     * }</pre>
+     *
+     * @implNote Implementations of this interface should provide their own
+     * implementation of this method.
+     *
+     * @see Gatherers
+     * @param <R> The element type of the new stream
+     * @param gatherer a gatherer
+     * @return the new stream
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    default <R> Stream<R> gather(Gatherer<? super T, ?, R> gatherer) {
+        return StreamSupport.stream(spliterator(), isParallel()).gather(gatherer);
+    }
+
```

Comment:
@briangoetz 

>A brief description of what a Gatherer is here (around the time you say "extension point") would be helpful, something general like "Gatherers are highly flexible, expressing a many-to-many transformation on the elements of a stream, and support short-circuiting, blah blah blah". ENough to make people want to read either the linked section or the Gatherer doc.

Good point. I went with:

`Gatherers are highly flexible and can describe a vast array of possibly stateful operations, with support for short-circuiting, and parallelization.`

>The @impleNote is a little unclear; I'd use the format for other ops added after 8, such as takeWhile or mapMulti.

Do you mean the implSpec? Because the implNote just states: "Implementations of this interface should provide their own implementation of this method."

>I'd add @see Gatherer.

:+1: